src/toctreeview.cpp

```diff
@@ -0,0 +1,182 @@
+#include "toctreeview.h"
+#include "latexstructure.h"
+#include "latexdocument.h"
+#include "configmanager.h"
+#include "utilsUI.h"
+#include "utilsSystem.h"
+#include "editors.h"
+#include "latexeditorview.h"
+#include "tocitemtree.h"
+#include "tocitemdata.h"
+
+TocTreeView::TocTreeView(const ConfigManager& config, LatexDocuments* docs, QWidget* parent) :
+	QTreeView(parent),
+	documents(docs)
+{
+	header()->hide();
+	setObjectName("TocTree");
+	UtilsUi::enableTouchScrolling(this);
+
+	if (config.indentationInStructure > 0)
+		setIndentation(config.indentationInStructure);
+
+	updateTOC();
+}
+
+void TocTreeView::updateTOC()
+{
+	closedDocs.clear();
+	updateData();
+	expandAll();
+}
+
+QList<TocItem> parseStruct(QString file, StructureEntry* str) {
+	QList<TocItem> result;
+	for (int i = 0; i < str->children.count(); i++) {
+		if (str->children[i]->type != StructureEntry::SE_SECTION && str->children[i]->type != StructureEntry::SE_INCLUDE) continue;
+		TocItem item;
+		item.file = file;
+		item.title = str->children[i]->title;
+		item.level = str->children[i]->level;
+		item.line = str->children[i]->getRealLineNumber();
+		if (str->children[i]->type == StructureEntry::SE_INCLUDE) {
+			if (!item.title.endsWith(".tex", Qt::CaseInsensitive)) item.title.append(".tex");
+			item.title = QFileInfo(file).dir().filePath(item.title);
+			item.level = -1;
+		}
+		if (!result.contains(item)) result.push_back(item);
+		if (str->children[i]->type == StructureEntry::SE_SECTION) result.append(parseStruct(file, str->children[i]));
+	}
+	return result;
+}
+
+struct QDataComparer
+{
+	bool operator()(const QList<QVariant>& a, const QList<QVariant>& b) const
+	{
+		return a[1].toInt() < b[1].toInt();
+	}
+};
+
+QList<TocItem> parseFile(QString filePath)
```

Comment:
there should not be a file parser. txs loads included files for complete analysis and that should do it.
The file parser is way less capable (not verbatim/comment detection, no adaptability for packages, etc.etc) and it will give a noticeable difference between document toc and project toc.